### PR TITLE
Shared Model: Amount construction

### PR DESCRIPTION
**Pull request by MBoldyrev**
### Description of the Change

This test covers various cases of Amount constructor input data.
Also the constructor itself was rewritten (preserving the old behaviour except the case of invalid input, in which the toStringRepr now provides not the original input, but "NaN").

### Benefits

Better test coverage. Prevent the random data from the network from going happily to the database through `toStringRepr()`.

### CODE REVIEW

File: shared_model/interfaces/common_objects/impl/amount.cpp [hunk 1]
```diff
@@ -9,39 +9,45 @@
...
-
-    Amount::Amount(const Amount &o) : Amount(std::string(o.amount_)) {}
+    Amount::Amount(const std::string &amount)
+        : string_repr_("NaN"), precision_(0), multiprecision_repr_(0) {
+      const auto dot_pos = amount.find_first_not_of(kDigits);
+      if (dot_pos != std::string::npos) {
+        // fail, if:
+        if (amount[dot_pos]
+                != kDecimalSeparator  // string contains an invalid character
+            or amount.find_first_not_of(kDigits, dot_pos + 1)
```
Comment (igor-egorov) [inline shared_model/interfaces/common_objects/impl/amount.cpp]:
```suggestion
            and amount.find_first_not_of(kDigits, dot_pos + 1)
```

File: shared_model/interfaces/common_objects/amount.hpp [hunk 2]
```diff
@@ -59,10 +55,9 @@ namespace shared_model {
       Amount *clone() const override;
 
      private:
-      const std::string amount_;
-
+      std::string string_repr_;
```
Comment (luckychess) [inline shared_model/interfaces/common_objects/amount.hpp]: it makes me nervous a bit that we actually have 2 different backend for a single object. It makes possible (in theory) desynchronisation between. Maybe just recalculate `string_repr_ ` on fly? I could have a negative performance impact but I don't think it's big.

Comment (MBoldyrev) [reply shared_model/interfaces/common_objects/amount.hpp]: I see. What if I make both representations const?

Comment (luckychess) [reply shared_model/interfaces/common_objects/amount.hpp]: It would be definitely better.

**Pull request merged**